github/release_notes/__init__.py

```diff
@@ -0,0 +1,353 @@
+
+def commits_in_range(
+    repo: git.Repo,
+    commit_range: str,
+    repository_branch: str=None
+):
+    args = [commit_range]
+    if repository_branch:
+        args.append(repository_branch)
+
+    GIT_FORMAT_KEYS = [
+        "%H",   # commit hash
+        "%s",   # subject
+        "%B"    # raw body
+    ]
+    pretty_format = '%x00'.join(GIT_FORMAT_KEYS) # field separator
+    pretty_format += '%x01' #line ending
+
+    kwargs = {'pretty': pretty_format}
+    git_logs = _.split(repo.git.log(*args, **kwargs), '\x01')
+
+    return commits_from_logs(git_logs)
+
+def commits_from_logs(
+    git_logs: list
+):
+    r = re.compile(
+        r"(?P<commit_hash>\S+?)\x00(?P<commit_subject>.*)\x00(?P<commit_message>.*)",
+        re.MULTILINE | re.DOTALL
+    )
+
+    commits = _\
+        .chain(git_logs) \
+        .map(lambda c: r.finditer(c)) \
+        .map(lambda iter: next(iter)) \
+        .map(lambda m: m.groupdict()) \
+        .map(lambda g: Commit(
+            hash=g['commit_hash'],
+            subject=g['commit_subject'],
+            message=g['commit_message']
+        )) \
+        .value()
+    return commits
+
+def fetch_pr_numbers_from_commits(
+    commits: list
+) -> set:
+    pr_numbers = set()
+    for commit in commits:
+        pr_number = pr_number_from_subject(commit.subject)
+
+        if pr_number:
+            pr_numbers.add(pr_number)
+
+    return pr_numbers
+
+def pr_number_from_subject(commit_subject: str):
+    pr_number = _.head(re.findall(r"Merge pull request #(\d+)", commit_subject))
+    if not pr_number: # Squash commit
+        pr_number = _.head(re.findall(r" \(#(\d+)\)$", commit_subject))
+    return pr_number
+
+def fetch_release_notes_from_prs(
+    helper: GitHubRepositoryHelper,
+    pr_numbers_in_range: set,
+    cn_current_repo: ComponentName
+) -> list:
+    # we should consider adding a release-note label to the PRs
+    # to reduce the number of search results
+    prs_iter = helper.search_issues_in_repo('type:pull is:closed')
+
+    release_notes = list()
+    for pr_iter in prs_iter:
+        pr_dict = pr_iter.as_dict()
+
+        pr_number = pr_dict['number']
+        if not str(pr_number) in pr_numbers_in_range:
+            continue
+
+        release_notes_pr = extract_release_notes(
+            reference_id=pr_number,
+            text=pr_dict['body'],
+            user_login=_.get(pr_dict, 'user.login'),
+            cn_current_repo=cn_current_repo,
+            reference_is_pr=True
+        )
+        if not release_notes_pr:
+            continue
+
+        release_notes.extend(release_notes_pr)
+    return release_notes
+
+def fetch_release_notes_from_commits(
+    commits: list,
+    cn_current_repo: ComponentName
+):
+    release_notes = list()
+    for commit in commits:
+
+        release_notes_commit = extract_release_notes(
+            reference_id=commit.hash,
+            text=commit.message,
+            user_login=None, # we do not have the gitHub user at hand
```

Comment:
the user_login str in this case is not used for any github access, hence no quota issue.
It is used for the release note text like
>Added new feature foo (#1, @\<user_login\>)

and for transporting the release notes via pull request, then it would look like

\`\`\` improvement operator github.com/foo/bar #1 @\<user_login\>
Added new feature foo
\`\`\`